libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/room/member/RoomMemberListFetcher.kt

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.matrix.impl.room.member
+
+import io.element.android.libraries.core.coroutine.parallelMap
+import io.element.android.libraries.matrix.api.room.MatrixRoomMembersState
+import io.element.android.libraries.matrix.api.room.roomMembers
+import kotlinx.collections.immutable.toImmutableList
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import org.matrix.rustcomponents.sdk.RoomInterface
+import org.matrix.rustcomponents.sdk.RoomMembersIterator
+import org.matrix.rustcomponents.sdk.use
+import timber.log.Timber
+import kotlin.coroutines.cancellation.CancellationException
+import kotlin.coroutines.coroutineContext
+
+/**
+ * This class fetches the room members for a given room in a 'paginated' way, and taking into account previous cached values.
+ */
+internal class RoomMemberListFetcher(
+    private val room: RoomInterface,
+    private val dispatcher: CoroutineDispatcher,
+    private val pageSize: Int = 1000,
+) {
+    private val updatedRoomMemberMutex = Mutex()
+    private val roomId = room.id()
+
+    private val _membersFlow = MutableStateFlow<MatrixRoomMembersState>(MatrixRoomMembersState.Unknown)
+    val membersFlow: StateFlow<MatrixRoomMembersState> = _membersFlow
+
+    /**
+     * Fetches the room members for the given room.
+     * It will emit the cached members first, and then the updated members in batches of [pageSize] items, through [membersFlow].
+     * @param withCache Whether to load the cached members first. Defaults to true.
+     */
+    suspend fun fetchRoomMembers(withCache: Boolean = true) {
+        if (updatedRoomMemberMutex.isLocked) {
+            Timber.i("Room members are already being updated for room $roomId")
+            return
+        }
+        updatedRoomMemberMutex.withLock {
+            withContext(dispatcher) {
+                // Load cached members as fallback and to get faster results
+                if (withCache) {
+                    if (_membersFlow.value !is MatrixRoomMembersState.Ready) {
+                        fetchCachedRoomMembers()
+                    } else {
+                        Timber.i("No need to load cached members found for room $roomId")
+                    }
+                }
+
+                val prevRoomMembers = (_membersFlow.value as? MatrixRoomMembersState.Ready)?.roomMembers?.toImmutableList()
+                _membersFlow.value = MatrixRoomMembersState.Pending(prevRoomMembers = prevRoomMembers)
+
+                try {
+                    // Start loading new members
+                    parseAndEmitMembers(room.members())
```

Comment:
In case of cancellation I think we can leak members here. Check how it was done in the previous implementation